### DISCUSSION TITLE
Seems like there is no Server for external files ?

### DISCUSSION THREAD

**Discussion by jacquesmignault** [httnn/nih-plug-webview#10] [Questions]
I am interested in your work on this nih webview demo.
I am a noobie at Rust and I’d like to find out how to do this...

I am trying to point line 134 of your [lib.rs](http://lib.rs/) to my own index.html from an angular app.
 let editor = WebViewEditor::new(HTMLSource::String(include_str!("dist/index.html")), (200, 200))…….

it is an angular app and here are the errors I’m getting:
![Screenshot 2023-07-12 at 9 06 02 AM](https://github.com/maxjvh/nih-plug-webview/assets/5486127/ab1fb01e-73b1-46e9-8aaa-7b016611c0d3)

It’s loading the index.html but it’s not getting the included files because “localhost” doesn’t seem active ?

I’m not the best rust programmer, but shouldn’t there be a web server (Tauri) underneath ?

**Comment by httnn** [comment]
hey!

this library isn't supposed to be a full web framework for plugins, but just the bare minimum needed to run a browser inside a plugin and communicate with it efficiently.

that being said, it should be pretty straightforward to use this library with a development server - you'll just have to start the server yourself. of course, when distributing your plugin you'll want to bundle the entire web app in a way that doesn't require a server.

this is what i'm doing in a project of mine currently:
```rs
#[cfg(debug_assertions)]
let src = HTMLSource::URL("http://localhost:3000");
#[cfg(not(debug_assertions))]
let src = HTMLSource::String(include_str!("../../ui/dist/index.html"));

let editor = WebViewEditor::new(src, (950, 780))
```
so in development builds it's loading the page running on localhost via the URL. but when building a release build (with cargo's `--release` argument), it will instead bundle and load a production-ready "index.html". you could use e.g. Vite and https://github.com/richardtallent/vite-plugin-singlefile to bundle _everything_ into that index.html file. that way it would contain everything needed for the page and no server is needed.

if you don't want to bundle everything into a single file in release builds, you could also use a custom protocol to load your assets:
```rs
static WEB_ASSETS: Dir<'_> = include_dir!("../ui/dist/assets");

...

.with_custom_protocol("webview".to_owned(), |req| {
    if let Some(file) = WEB_ASSETS.get_file(req.uri().path().trim_start_matches("/")) {
        return Response::builder()
            .header(
                "content-type",
                match file.path().extension().unwrap().to_str().unwrap() {
                    "js" => "text/javascript",
                    "css" => "text/css",
                    "ttf" => "font/ttf",
                    _ => "",
                },
            )
            .header("Access-Control-Allow-Origin", "*")
            .body(file.contents().into())
            .map_err(Into::into);
    }
    panic!("Web asset not found.")
})
...
```

with this code, the closure passed to `with_custom_protocol` will handle all requests sent with the `webview://` protocol.

anyway, if you just want to play around you could just use development server and worry about release builds later:
```rs
let editor = WebViewEditor::new(HTMLSource::URL("http://localhost:3000"), (950, 780))
```

hope this helps!
